### PR TITLE
Hive metastore integration with installation files and a sample application

### PR DESCRIPTION
**Pull request by srikumar003**

### CODE REVIEW

File: release-tools/Makefile [hunk 1]
```diff
@@ -102,7 +102,7 @@ ifeq ($(PULL_COMPONENTS), true)
 	docker pull $(DATASET_OPERATOR_IMAGE)
 else
 	$(call install_local,csi-s3,$(CSI_S3_IMAGE))
-	$(call install_local,csi-driver-nfs,$(CSI_NFS_IMAGE))
+#	$(call install_local,csi-driver-nfs,$(CSI_NFS_IMAGE))
```
Comment (YiannisGkoufas) [inline release-tools/Makefile]: Why is this commented out? it doesn't build the nfs driver. if I comment it out it fails to build.
```
go: cannot determine module path for source directory /csi-driver-nfs (outside GOPATH, no import comments)
release-tools/build.make:64: recipe for target 'build-nfsplugin' failed
make: *** [build-nfsplugin] Error 1
make: *** [build-containers] Error 2
```

File: src/dataset-operator/pkg/controller/dataset/dataset_controller.go [hunk 2]
```diff
@@ -428,59 +428,95 @@ func processRemoteDataset(cr *comv1alpha1.Dataset, rc *ReconcileDataset) (reconc
...
+		if err = controllerutil.SetControllerReference(cr, configMapObject, rc.scheme); err != nil {
+			return result, err
+		}
+
+		foundConfigMap := &corev1.ConfigMap{}
+		err = rc.client.Get(context.TODO(), types.NamespacedName{Name: configMapObject.Name, Namespace: configMapObject.Namespace}, foundConfigMap)
 
-			if err = controllerutil.SetControllerReference(cr, configMapObject, rc.scheme); err != nil {
+		if err != nil && errors.IsNotFound(err) {
+			processRemoteDatasetLogger.Info("Creating new configMap", "configMap.namespace",
```
[the remark this answers is not in the document]
Comment (srikumar003) [reply src/dataset-operator/pkg/controller/dataset/dataset_controller.go]: Fixed it in https://github.com/srikumar003/dataset-lifecycle-framework/commit/6a9b2abb10263fd88841c452c938fdaa5086d79b

File: Makefile [hunk 3]
```diff
@@ -1,5 +1,5 @@
-DOCKER_REGISTRY_COMPONENTS := the_registry_to_use_for_components
-DOCKER_REGISTRY_SECRET := your_already_installed_secrets
+DOCKER_REGISTRY_COMPONENTS := res-drl-hpc-docker-local.artifactory.swg-devops.com
+DOCKER_REGISTRY_SECRET := art-drl-hpc
```
Comment (christian-pinto) [inline Makefile]: please revert these changes

File: Makefile [hunk 4]
```diff
@@ -8,8 +8,8 @@ DOCKER_REGISTRY_SIDECARS := quay.io/k8scsi
 # if you pull from public use quay.io/k8scsi
 PULL_SIDECARS := true
 
-DATASET_OPERATOR_NAMESPACE := default
-NAMESPACES_TO_MONITOR := default
+DATASET_OPERATOR_NAMESPACE := drlmcdlf
+NAMESPACES_TO_MONITOR := drlmcdlf
```
Comment (christian-pinto) [inline Makefile]: And these too :smiley:

File: src/dataset-operator/pkg/admissioncontroller/mutate.go [hunk 5]
```diff
@@ -127,53 +132,49 @@ func Mutate(body []byte) ([]byte, error) {
...
-						},
-					}
-
-					values = append(values, configmap_ref)
-					values = append(values, secret_ref)
+			for _, config_toinject := range configs_toinject {
+				//TODO: Check if the configmap reference exists in the API server
+
+				configmap_ref := map[string]interface{}{
+					"prefix": config_toinject,
```
Comment (christian-pinto) [inline src/dataset-operator/pkg/admissioncontroller/mutate.go]: maybe add a "-" after the prefix otherwise env variables names look weird such as **example-datasetbucket**. Please do the same with the secret

File: src/dataset-operator/pkg/controller/dataset/dataset_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (christian-pinto) [reply src/dataset-operator/pkg/controller/dataset/dataset_controller.go]: yep, fix works. Just a minor stylish thing to be added, added a comment for this.

File: examples/hive/k8s/README.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+should be installed in the same namespace as the rest of DLF. This
+guide assumes Minikube as the target Kubernetes cluster, but it is applicable to any Kubernetes/Openshift infrastructure.
+
+## Initial steps
+
+First, some configuration. The ObjectStorage integration of Hive requires that the endpoint be provided at the point of initial configuration. 
+if you are using the Minio install as described in the main installation guide, the following applies.
+
+```
+$ export S3_ENDPOINT=$(minikube service minio-service --url)
```
[the remark this answers is not in the document]
Comment (srikumar003) [reply examples/hive/k8s/README.md]: Changed Hive Readme to use Nooba b1ceb62099b1b67a088c682d361c3bc3a16c3896

File: examples/hive/k8s/README.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+$ export S£_ENDPOINT=https://<Object_Storage_Service_Provider_URL>
+```
+```
+$ envsubst < conf/hive-site.tmpl | tee conf/hive-site.xml
+```
+```
+$ envsubst < conf/metastore-site.tmpl | tee conf/metastore-site.xml
+```
+```
+$ mv deploy/s3-secret.tmpl deploy/s3-secret.yaml
```
[the remark this answers is not in the document]
Comment (srikumar003) [reply examples/hive/k8s/README.md]: No need for this file 2fb204cc047990efb1406b068409e8952bca34c9

File: examples/hive/k8s/README.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+$ kubectl apply -f deploy/s3-secret.yaml
```
[the remark this answers is not in the document]
Comment (srikumar003) [reply examples/hive/k8s/README.md]: Updated Readme to use these variables b1ceb62099b1b67a088c682d361c3bc3a16c3896

File: examples/hive/k8s/Makefile [hunk 9]
```diff
@@ -57,7 +65,11 @@ push-images: build-images
 	@docker push $(HIVESERVER_IMAGE) ;\
 	docker push $(HIVEMETASTORE_IMAGE)
 
-deploy-hive: push-images 
+minikube-load-containers: build-images
+	$(call load_containers_minikube, $(HIVEMETASTORE_IMAGE), hivemetastore)
+	$(call load_containers_minikube, $(HIVESERVER_IMAGE), hiveserver)
+
+deploy-hive: push-images
```
[the remark this answers is not in the document]
Comment (srikumar003) [reply examples/hive/k8s/Makefile]: Fixed this in 9c66ac156452e7d4c3fd366a85116342b694b03f

File: examples/hive/k8s/Makefile [hunk 10]
```diff
@@ -32,6 +34,12 @@ MAKE_ENV += HIVEMETASTORE_IMAGE
 SHELL_EXPORT := $(foreach v,$(MAKE_ENV),$(v)='$($(v))' )
 
 #K8S_FILES += $(shell find $(HIVE_FILE_PATH)/deploy -maxdepth 1 -name '*.yaml')
+define load_containers_minikube
+    @docker save $(2) | gzip > _tmp/$(1).tar.gz;\
```
[the remark this answers is not in the document]
Comment (srikumar003) [reply examples/hive/k8s/Makefile]: Fixed this in 9c66ac156452e7d4c3fd366a85116342b694b03f

File: examples/hive/k8s/Makefile [hunk 11]
```diff
@@ -57,7 +65,11 @@ push-images: build-images
 	@docker push $(HIVESERVER_IMAGE) ;\
 	docker push $(HIVEMETASTORE_IMAGE)
 
-deploy-hive: push-images 
+minikube-load-containers: build-images
+	$(call load_containers_minikube, $(HIVEMETASTORE_IMAGE), hivemetastore)
+	$(call load_containers_minikube, $(HIVESERVER_IMAGE), hiveserver)
```
[the remark this answers is not in the document]
Comment (srikumar003) [reply examples/hive/k8s/Makefile]: Swapped the order of parameters 1 and 2 in the function 9c66ac156452e7d4c3fd366a85116342b694b03f

File: Makefile (hunk 3 above)
Comment (srikumar003) [reply Makefile]: Fixed this in 9c66ac156452e7d4c3fd366a85116342b694b03f

File: Makefile (hunk 4 above)
Comment (srikumar003) [reply Makefile]: fixed 9c66ac156452e7d4c3fd366a85116342b694b03f

File: examples/hive/k8s/deploy/hiveserver.yaml [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+      app: hiveserver
+  template:
+    metadata:
+      labels:
+        app: hiveserver
+    spec:
+      containers:
+        - name: hiveserver-container
+          image: ${HIVESERVER_IMAGE}
+          imagePullPolicy: "Always"
```
Comment (christian-pinto) [inline examples/hive/k8s/deploy/hiveserver.yaml]: Change pull policy, it fails in minikube because it tries pulling the image. **IfNotPresent** should make the trick

File: examples/hive/k8s/deploy/metastore.yaml [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+      app: hivemetastore
+  template:
+    metadata:
+      labels:
+        app: hivemetastore
+    spec:
+      containers:
+        - name: hivemetastore-container
+          image: ${HIVEMETASTORE_IMAGE}
+          imagePullPolicy: "Always"
```
Comment (christian-pinto) [inline examples/hive/k8s/deploy/metastore.yaml]: Change pull policy, it fails in minikube because it tries pulling the image. **IfNotPresent** should make the trick

File: examples/hive/sampleapp/populate.sh [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+HIVE_CLI_PORT=`kubectl get svc hiveserver -n default -o jsonpath='{.spec.ports[?(@.name=="cliservice")].nodePort}'`
+
+HIVE_CLI_IP=`minikube service hiveserver --url -n default | awk -F':' -v port="$HIVE_CLI_PORT" '{if ($3 == port) print $2}' - | cut -d / -f 3`
+
+echo $HIVE_CLI_IP:$HIVE_CLI_PORT
+
+DOCKER_REGISTRY_COMPONENTS="registry_used_in_installation"
+HIVESERVER_IMAGE="hive-server:latest"
+
+docker run -v $(PWD):/sampleapp -it ${DOCKER_REGISTRY_COMPONENTS}/${HIVESERVER_IMAGE} bin/beeline -u "jdbc:hive2://$HIVE_CLI_IP:$HIVE_CLI_PORT/;transportMode=http;httpPath=/cliservice" -f /sampleapp/sample.hql
```
Comment (christian-pinto) [inline examples/hive/sampleapp/populate.sh]: $(PWD) -> ${PWD}

Comment (christian-pinto) [reply examples/hive/sampleapp/populate.sh]: Also, this last command returns:
```
20/03/13 09:26:22 [main]: WARN jdbc.HiveConnection: Failed to connect to 192.168.39.51:31321
Error: Could not open client transport with JDBC Uri: jdbc:hive2://192.168.39.51:31321/;transportMode=http;httpPath=/cliservice: Could not establish connection to jdbc:hive2://192.168.39.51:31321/;transportMode=http;httpPath=/cliservice: org.apache.http.conn.HttpHostConnectException: Connect to 192.168.39.51:31321 [/192.168.39.51] failed: Connection refused (Connection refused) (state=08S01,code=0)

```

File: examples/hive/sampleapp/README.md [hunk 15]
```diff
@@ -1,23 +1,45 @@
...
 ```
-$ mc mb local/book-test
+$ s3cmd mb s3://book-test  
 ```
 
 2. Upload the given CSV file to the bucket 
 
 ```
-$ mc cp books.csv local/book-test
+$ s3cmd put books.csv s3://book-test/
```
Comment (christian-pinto) [inline examples/hive/sampleapp/README.md]: All this could be automated the same way we load example data in NooBaa for testing the regular datasets. There is a docker image wtih _awscli_ that is built during the installation process and you could use that one to load the csv inside the new bucket.

See [data-loader-noobaa.yaml](https://github.com/IBM/dataset-lifecycle-framework/blob/master/examples/noobaa/data-loader-noobaa.yaml) and [noobaa_install.sh](https://github.com/IBM/dataset-lifecycle-framework/blob/master/examples/noobaa/noobaa_install.sh) functions `build_data_loader` and `run_data_loader`.

**Comment by YiannisGkoufas** [comment]
I am probably configuring something wrong in the hive example on the `s3cmd --configure` step.
While this `s3cmd ls` works and shows the buckets:
```
2020-03-23 10:48  s3://first.bucket
2020-03-23 10:48  s3://my-bucket-5fc7f163-aeba-459b-8b66-69caeeec3887
```
However, when I try to do this `s3cmd mb s3://book-test` I get the error:
`ERROR: S3 error: 403 (InvalidAccessKeyId): The AWS Access Key Id you provided does not exist in our records.`

As I understand the `s3cmd --configure` command creates a `~/.s3cfg` file which stores the configuration parameters. Maybe showing an example of `~/.s3cfg` might be helpful

File: examples/hive/sampleapp/populate.sh (hunk 14 above)
Comment (srikumar003) [reply examples/hive/sampleapp/populate.sh]: Variable substitution fixed in https://github.com/srikumar003/dataset-lifecycle-framework/commit/74d2c763bc1bdb3190430c2ea89f60c7c6cfc999

Provided a convenience script for setting up sample app

File: examples/hive/k8s/README.md [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+```
+```
+$ kubectl apply -f deploy/postgres-secret.yaml
```
[the remark this answers is not in the document]
Comment (srikumar003) [reply examples/hive/k8s/README.md]: The Makefile for hive installation now takes care of all the configuration file modification. Fixed in https://github.com/srikumar003/dataset-lifecycle-framework/commit/2f40baf08e9f33f466e8c10283cf3cf4ce885ac4

File: examples/hive/sampleapp/README.md (hunk 15 above)
Comment (srikumar003) [reply examples/hive/sampleapp/README.md]: Provided a convenience script for creating the sample application: https://github.com/srikumar003/dataset-lifecycle-framework/commit/ae9a885eed1c985fabbad5155531a2ea7734cb03

**Comment by srikumar003** [comment]
> I am probably configuring something wrong in the hive example on the `s3cmd --configure` step.
> While this `s3cmd ls` works and shows the buckets:
> 
> ```
> 2020-03-23 10:48  s3://first.bucket
> 2020-03-23 10:48  s3://my-bucket-5fc7f163-aeba-459b-8b66-69caeeec3887
> ```
> 
> However, when I try to do this `s3cmd mb s3://book-test` I get the error:
> `ERROR: S3 error: 403 (InvalidAccessKeyId): The AWS Access Key Id you provided does not exist in our records.`
> 
> As I understand the `s3cmd --configure` command creates a `~/.s3cfg` file which stores the configuration parameters. Maybe showing an example of `~/.s3cfg` might be helpful

Removed the need for configuring `s3cmd`. We are using the Docker image for awscli that is wrapped in a convenience script to set up the sample application. Fixed in https://github.com/srikumar003/dataset-lifecycle-framework/commit/ae9a885eed1c985fabbad5155531a2ea7734cb03

**Comment by christian-pinto** [comment]
There is still an issue with the main makefile. Your branch contains the wrong values for registry, secret  and namespace that make the process fail while installing on minikube.

**Comment by christian-pinto** [comment]
Looks good now!
